CONTRIBUTORS.md

```diff
@@ -154,6 +154,93 @@ All contributions should use the same whitespacing as the rest of the project.
 Moreover, Pull requests where a large number of changes only deal with whitespace will be
 rejected.
 
+# Comments
+Comments are very important. Not only are they are an aid to the reader, but in many
+cases the comments specify the contract against which the code is evaluated.
+Without a clear contract for a function, for example, it is *impossible* to determine
+whether the function is implemented correclty.
+
+The following kinds of things should have comments.
+
+## Components (`struct`'s and `trait`'s)
+Comments for a component (`struct` or `trait`) should explain what the overall
+purpose of that component is. This is usually a concept, and not a formal contract.
+
+Help the reader understand why this group of data or methods are grouped
+together. By explaining what a component *does* we also implicitly explain what
+it *does not do* in a modular architecture.
+
+Example:
+
+```
+/// The `BurnStateDB` allows the user to query the state of the underlying
+/// "burn chain", which in practice is the Bitcoin chain.
+trait BurnChainDB {
+```
+
+## Functions
+The comments on a function should explain the contract that the function
+implements. This is very important. Without a clear contract, we cannot
+evaluate the correctness of the implementation.
+
+Without being unnecessarily verbose, explain how the output is calculated
+from the inputs. Explain any restrictions on the inputs. Explain error
+conditions, including when the function will panic, return an error
+or return an empty value.
+
+Example:
+
+```
+pub trait BurnStateDB {
+    /// Returns the *header hash* of the burnchain block at the
+    /// burnchain height `height` in sortition `sortition_id`.
+    ///
+    /// Returns None if there is no block at this height.
+    fn get_burn_header_hash(
+        &self,
+        height: u32,
+        sortition_id: &SortitionId,
+    ) -> Option<BurnchainHeaderHash>;
+```
```

Comment:
Do we want to give an example of a function with a markdown subheading (like with a panics section?)

Reply:
I can add one or you can make some revisions.

Reply:
I am just adding the example from my pr here as another example for documentation for a function